### ISSUE TITLE
How to use the Drush driver with the Behat Drush endpoint to create nodes?

### ISSUE THREAD

**Issue by dsnopek** [jhedstrom/DrupalDriver#90]
Per this blog post:

https://pantheon.io/blog/better-behavior-driven-development-remote-servers

And the merged PR:

https://github.com/jhedstrom/DrupalDriver/pull/56

It sounds like this should be possible...

But how to get this to actually work? I've tried setting up a Behat enviroment (with dev-master of both drupal/drupal-extension and drupal/drupal-driver) that registers the 'drush' driver, but not the 'drupal' driver, and every time I try to use a node creation step (from the DrupalContext - not DrushContent, which doesn't have any steps for creating nodes) it gives me an exception with the message: `Driver "drupal" is not registered`.

Do I need to make a custom context with node creation steps that use the underlying Drush driver directly? Will the DrupalContext be updated at some point to use the Drush driver for node creation, or will there be a new context added?

Thanks for your help!

**Comment by jhedstrom** [comment]
> `Driver "drupal" is not registered`
I think this means that somewhere in your `behat.yml` *or* custom code (or both), there is still reference to the `drupal` (eg, the Drupal API) driver whereas the goal here is to use drush.

Running behat with the verbose flag may provide more insight. Also, if it still isn't clear what's going on, perhaps you could post the `behat.yml` file you're using?

@greg-1-anderson might have some insights here too.

**Comment by greg-1-anderson** [comment]
Are you using Drupal 7, or Drupal 8?  I have a project for Drupal 7 here:

https://github.com/westkingdom/website/blob/master/behat.yml

However, when I switched this over to use the `drush` driver, I got:
```
No ability to create terms in Drupal\Driver\DrushDriver. Put `@api` into your feature and add an API driver (ex: `api_driver: drupal`) in behat.yml.(Drupal\Driver\Exception\UnsupportedDriverActionException)
```

At one time, I submitted a PR that added the ability to create nodes to the Drush driver, and I thought this was merged; however, it appears to not be there now. I don't know if that means it never made it into an official release, or if it was accidentally backed out.  I'll have to investigate.

**Comment by jhedstrom** [comment]
I might need to cut a new release now that you mention it. I think all this work was targeted for  the `1.2.0` milestone.

**Comment by greg-1-anderson** [comment]
Here's the PR that was merged:

https://github.com/jhedstrom/DrupalDriver/pull/56

**Comment by greg-1-anderson** [comment]
@dsnopek: You said that you are using dev-master of the DrupalDriver, which appears to still have the code from #56. You also said that the DrushContext had no steps for creating nodes, but it does (they are inherited from the parent class). Maybe you just missed this?

I did the following in my project:

composer require drupal/drupal-driver:dev-master

This brought the appropriate code in to my Drush Driver in my vendor directory; however, running my tests then came up with different errors.  I will investigate; seems no one has been using this code for a while, and it seems to have broken.

**Comment by dsnopek** [comment]
Thanks @greg-1-anderson and @jhedstrom!

I figured out where I was going wrong - I needed to set `api_driver: drush` - for some reason, I was assuming that by giving settings for 'drush' and not 'drupal' it would just choose 'drush' but it was still using 'drupal'.

```
You also said that the DrushContext had no steps for creating nodes, but it does (they are inherited from the parent class). Maybe you just missed this?
```

Actually, it doesn't! Its base class is RawDrupalContext which doesn't define any steps (like all the Raw* classes). But it looks like its the DrupalContext that I want to use, just with 'api_driver: drush'.

So, in summary, it was user error. :-) Thanks for checking this out! (... but sorry for distracting you with something was my mistake :-/)

**Comment by greg-1-anderson** [comment]
No problem at all; I suspect there are still bugs here.  Let me know if this starts working for you.

**Comment by dsnopek** [comment]
So, I've gotten further! I can use the behat-drush-endpoint to create nodes via Behat on a local site. But it's failing for an Drush alias on Pantheon.

Here's a gist with the full stack trace:

https://gist.github.com/dsnopek/d01b7897ad559eed37d85db1312e9d23

I tried to test running it directly from Drush and this is what I get:

```
$ drush @pantheon.openucb-1231-drs.dev behat create-node '{"type":"openberkeley_content_page","title":"New node via Drush"}'
PTY allocation request failed on channel 0
Command behat needs a higher bootstrap level to run - you will need      [error]
to invoke drush from a more functional Drupal environment to run this
command.
The drush command 'behat create-node                                     [error]                                                                                                                                                                                                                                             {"type":"openberkeley_content_page","title":"New node via Drush"}'
could not be executed.
Connection to appserver.dev.8e025b42-012a-4e50-b499-e8a18507eacc.drush.in closed.
```

If I run `drush @pantheon.openucb-1231-drs.dev help` I see the 'behat' command in there, so it's recognizing it, but just having some problem bootstrapping Drupal? Or maybe I'm running my test command wrong?

Is there any special configuration I need for running on Pantheon or an actually remote server?

**Comment by dsnopek** [comment]
Heh, hang on - don't do anything yet - I think I might have done something stupid :-)

**Comment by dsnopek** [comment]
Yes, ok, my fault again! All is working for me now :-)

**Comment by dsnopek** [comment]
Or, almost.. Running `drush behat create-node '...'` is working with Pantheon and creating the node, but it's still not working via Behat, even though it works fine with a local site..

**Comment by greg-1-anderson** [comment]
I was getting some errors in Behat as well. Seems things may have broken here.

**Comment by greg-1-anderson** [comment]
Are you working on Drupal 7 or Drupal 8? I think there are still errors in the Behat Drush Extension too. Maybe everything but Taxonomy working?  Don't quite remember.

**Comment by dsnopek** [comment]
Drupal 7 and Drush 7. Like I said, Behat works with a Drush alias that's on my local machine, just not an remote alias on Pantheon, which is pretty weird - it's the same behat.yml, I just change the alias. Running `drush behat create-node` works fine either locally or remotely on Pantheon.

The gist with the output I gave earlier is still what I see:

https://gist.github.com/dsnopek/d01b7897ad559eed37d85db1312e9d23

**Comment by dsnopek** [comment]
Ooo, a fun oddity of remote aliases with drush is that it never writes to the error stream (due to lots of `2>&1`) so the line at DrushDriver.php:298:

```
throw new \RuntimeException($process->getErrorOutput());
```

... will never show anything useful for a remote alias. But if I change that to:

```
throw new \RuntimeException($process->getOutput() . $process->getErrorOutput());
```

Then I get this message:

```
Pseudo-terminal will not be allocated because stdin is not a terminal.
Command not supported as typed.
```

Any ideas?

**Comment by greg-1-anderson** [comment]
Yep. "Command not supported as typed" means that pantheonssh is rejecting your ssh command, because it includes prohibited characters.  ';' is one example of something that is not allowed. I remember submitting a PR or two to adjust the characters used in random strings to make this work previously.

Can you dump the args being sent to the Process() class so that we can see what's in there for your test?

**Comment by greg-1-anderson** [comment]
There are only a few prohibited characters:
```
   ;
   $
   !
   `
```

**Comment by dsnopek** [comment]
Heh, ok, here's the command it's trying to execute:

```
drush @pantheon.openucb-1231-drs.dev  --nocolor=1  behat create-node '{"title":"Stock photo test","type":"openberkeley_content_page","body":"NUIa\/j=O~_fXrcYl+d9nn6c*2'\''\/&#(i[QB8;`3kQ?Jsq?it:`n+\"\\GCd<\/BkCB5iua<kk?nn7dkQ<W+7mT9][&S_e|ym'\''t>Vt~.;ZA\"yI{=ErYLkRcztAp-R|oqwY!][qlyc^Tp hL6%UD$RYNiprz:5Pqr*AIF[NF\/x,HC>9G};k\"+1K{qjgcq]aUeDoj k8(Mx=,|I`.Y_Lfo&|A2b`[LvuQ]Fme-Le>{i X2ZZ*X):'\''C0T.f5<3h?c5`Z,*|"}'
```

It's got to be that random string that it's using for the body. I'm using the Behat step:

```
Given I am viewing an "openberkeley_content_page" with the title "Stock photo test"
```

... which goes back to `DrupalContext->createNode()` which is using `$this->getRandom()->string(255)` to generate the body.

Maybe Behat should generate a random string that's more word-y with less symbols?

**Comment by greg-1-anderson** [comment]
Yeah, there is another method called `name` that is less-randomy that can be used. Thought I worked through these, and decreased the entropy used in create node, user, taxonomy term, etc.

It's possible that someone else came along and decided to increase entropy in order to "find more bugs".  Haven't looked into the history to see.  Anyway, you could try changing `string` to `name` in your local copy, and see if that fixes the test. If it does, maybe submit a PR.

A while back I had the idea that Drush should be more aggressive about putting the commandline arguments in stdin in backend mode. Started on it, but ran into some issues and set it aside. Don't remember what they were now.

**Comment by dsnopek** [comment]
Changing that to ```$this->getRandom()->name(255)``` fixes that first error! And, actually, I've been switching between dev-master and the last release a bunch, and apparently I was last on the release -- you already made that change in dev-master. :-) So, I'm back on dev-master...

But I'm now getting another error:

```
Catchable Fatal Error: Argument 2 passed to Drupal\DrupalExtension\Context\RawDrupalContext::dispatchHooks() must be an instance of stdClass, null given, called in vendor/drupal/drupal-extension/src/Drupal/DrupalExtension/Context/RawDrupalContext.php on line 316 and defined in vendor/drupal/drupal-extension/src/Drupal/DrupalExtension/Context/RawDrupalContext.php line 292
```

Apparently, `DrushDriver->createNode()` is returning NULL, rather than an object made by decoding the the JSON object returned from drush command.

Here's the output from Drush that it's trying to `json_decode()`:

```
Pseudo-terminal will not be allocated because stdin is not a terminal.
Invalid argument supplied for foreach() DefaultHandler.php:15          [warning]
{
    "title": "Stock photo test",
    "type": "openberkeley_content_page",
    "body": [],
    "status": 1,
    "promote": 0,
    "sticky": 0,
    "uid": "0",
    "created": 1465472894,
    "revision": false,
    "menu": {
        "link_title": "",
        "mlid": 0,
        "plid": 0,
        "menu_name": "main-menu",
        "weight": 0,
        "options": [],
        "module": "menu",
        "expanded": 0,
        "hidden": 0,
        "has_children": 0,
        "customized": 0,
        "parent_depth_limit": 8
    },
    "changed": 1465472894,
    "timestamp": 1465472894,
    "uuid": "9bcb49df-a20c-423b-a5e0-f41dc1375e41",
    "vuuid": "2c8f047f-e025-4374-a42b-2460734d8015",
    "log": "",
    "nid": "12",
    "language": "",
    "comment": 0,
    "tnid": 0,
    "translate": 0,
    "vid": "12"
}
```

Notice the two lines at the top. Those really should have gone to stderr, but because this is a remote alias and everything gets put into stdout (due some probably over-zealous `2>&1` action). I think this is making for invalid JSON. We should probably use a regex to remove everything before the first `{` character?

**Comment by dsnopek** [comment]
Huzzah! Here's a PR that finally gets this working for me:

https://github.com/jhedstrom/DrupalDriver/pull/91

Thanks for sticking with me (even though 90% of the problems were my fault ;-))

**Comment by jhedstrom** [comment]
Closing out. Thanks for all the debugging and work on this!
